TheThingsNetwork/lorawan-stack
# Introduce batch rights assertion for gateways

#### Summary



Introduce batch rights assertion for gateways. Refs: https://github.com/TheThingsIndustries/product-management/issues/6

#### Changes



- Add new `AssertRights` RPC for a new `GatewayBatchAccess` service.
- Add implementation and tests.
- Adapt `BatchGetGatewayConnectionStats` to use this RPC.
- Testing

#### Testing



1. Create one or two test gateways.
2. Check their stats using `ttn-lw-cli gtw get-connection-stats <gtw-id-1> <gtw-id-2>...`
3. Now connect these gateways 
4. Check stats again. Their connection stats should appear.
5. Now repeat with gateways where the user only has stats access.
6. Now check gateways where the user is not a collaborator but the status is public.
7. Now check that the call fails when a gateway ID that's a user is not a collaborator to is requested.

##### Regressions



Tests above should cover the regressions.

#### Notes for Reviewers



Motivation behind the RPC implementation.

- Rights check currently is implemented via the `ListRights` RPCs for entities.
- This RPC is wrapped by the `rights.RequireXXX` functions with add a cache, fetch rights from the context or check the component fetcher. The latter has different implementations for the IS and other components.
- For this particular use case (batch checks), I implemented a direct RPC for components to call for the following reasons 
   - Unlike the `ListRights` RPC, it's sufficient for this use case to return a yes/no for the entire batch. There is no need to fetch rights from the IS for each entity and compare that. As a result, nothing really needs to be cached.
   - Also, it's quite difficult to cache a call with multiple IDs.
   - This is not part of a hot path and is never intended to be used as such. 

#### Checklist
